### ISSUE TITLE
locals are not retained for the whole document

### ISSUE THREAD

**Issue by dead10ck** [helix-editor/helix#1151]
I reported this in the Matrix chat room, but filing an issue to track it.

I tried to improve the C highlighting to highlight macro invocations differently than regular function calls. After reading through the tree-sitter docs, it seems like [local variables](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables) should be able to enable this.

So far, I think I've followed the docs pretty closely and came up with [this](https://github.com/dead10ck/helix/commit/1e3813ae5211042b8efedcaef2aabd6534694b3e). However, it doesn't seem to be working. Macro invocations are still being colored like any other function call.

I figured out that locals do seem to work, but only if the definition is on screen.

[![asciicast](https://asciinema.org/a/TtsxTPTdiy7ztbOIVeUSntuKD.svg)](https://asciinema.org/a/TtsxTPTdiy7ztbOIVeUSntuKD)

As soon as you scroll past the definition, it loses it. I can see with added debug logging around [here](https://github.com/tree-sitter/tree-sitter/blob/e85a279cf29da1b08648e27214dda20a841e57c8/highlight/src/lib.rs#L791) that the `scope.local_defs` is at first populated with the macro definition, but when it goes out of sight, `scope.local_defs` becomes empty.

@archseer mentioned that it could be because highlights are done only for the range on the screen, but that locals should be kept for the whole document.

https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/editor.rs#L136-L142

### Reproduction steps

Define a language locals query that finds variable definitions, and color them differently. If the definition and the reference are both on screen, it works, but as soon as the definition is off screen, the highlight reverts back to its own tree-sitter match instead of deferring to the definition's highlighter.

### Environment

- Platform: Linux
- Helix version: v0.5.0-143-g8ae6d96

**Comment by dead10ck** [comment]
@archseer unfortunately it doesn't look like the recent incremental changes mentioned in #1378 have fixed this issue. [This](https://github.com/dead10ck/helix/commit/b26addbaea65ae77a5229121fc54f6781ce4b172) is the branch I am using to test locals used in highlighting.

**Comment by archseer** [comment]
Yeah I haven't addressed this yet, it requires some more thought. It's possible to resolve locals for the whole document when we deal with injections (but that's slightly inefficient if we don't need the whole set), but that doesn't solve the whole problem: we'd also need to know what highlighting query matches for that `local.definition` so we can use it for the reference.

A workaround would be resolving all the highlights ahead of time (so no querying at all happens during render), but that can use _a lot_ of memory on large documents.

We could include more styling information on the `locals.scm` queries themselves: `@local.definition.variable` would use `@variable` for the `@local.reference`

**Comment by dead10ck** [comment]
If we're worried about pathologically large documents, maybe we could do something kind of dumb: remember all definitions and highlight matches up to a certain size limit. If we reach the limit, bail out and don't remember anything (current behavior) so that we don't get erratic highlighting. This will enable local highlighting for 95% of cases, and fall back in extreme cases to save hx from blowing up.

**Comment by dead10ck** [comment]
Another thought: do we have to remember *all* references? Could we get away with querying all of the references that are contained in the *visible* scopes that lie in the view? What I mean by that is e.g.:

```c
int global = 1;

void foo() {
    char b;
}

// We only see from here down
void bar() {
    float f;
}
...
```

If our view starts at `bar`, then the scopes that lie within the view are `bar` and the top level scope. So the definitions we query for should be `bar`, `f`, and `global`.

**Comment by archseer** [comment]
Yes, it hasn't been addressed yet which is why this issue is still open.

**Comment by MatrixManAtYrService** [comment]
Here's a case where it's not about where the viewing window is.  Maybe it's a separate thing?  It seems very similar.

There's some fixed character count beyond which the highlighting behavior changes.  If your screen is small you might have to zoom out to see, but the color of the code at the bottom is sensitive to edits in the code above in ways that I think are surprising.


![641981](https://github.com/helix-editor/helix/assets/5834582/d34a9403-0917-4748-9f70-9e346b270ecf)

I realize that there is a syntax error in this code, so maybe the real issue is that the breakage takes 50 lines to show up, rather than showing up right away. 

[the code](https://gist.github.com/MatrixManAtYrService/43f794fbf3a2c7492aaff2c815487871)

```
helix 23.10 (f6021dd0)
```
